java/compilers/CFG.java

```diff
@@ -0,0 +1,189 @@
+package compilers;
+
+import compilers.util.Tuple;
+import org.apache.commons.collections4.multimap.ArrayListValuedHashMap;
+
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+class CFG {
+    private ArrayListValuedHashMap<Symbol, LinkedList<Symbol>> map;
+
+    CFG(ArrayListValuedHashMap<Symbol, LinkedList<Symbol>> map) {
+        this.map = map;
+    }
+
+    //<editor-fold desc="Getters and Things-that-return-streams">
+    Stream<Rule> productions() {
+        return map.entries().stream()
+                .map(e -> new Rule(e.getKey(), e.getValue()));
+    }
+
+    Stream<Rule> productions(Symbol lhs) {
+        return productions().filter(rule -> rule.getLeft().equals(lhs));
+    }
+
+    Stream<Symbol> terminals() {
+        return productions()
+                .flatMap(Rule::symbols)
+                .filter(Symbol::isTerminal)
+                .distinct();
+    }
+
+    Stream<Symbol> nonterminals() {
+        return productions()
+                .flatMap(Rule::symbols)
+                .filter(Symbol::isNonTerminal)
+                .distinct();
+    }
+
+    List<Rule> getProductions() {
+        return productions().collect(Collectors.toList());
+    }
+
+    List<Rule> getProductions(Symbol lhs) {
+        return productions(lhs).collect(Collectors.toList());
+    }
+
+    Set<Symbol> getTerminals() {
+        return terminals().collect(Collectors.toSet());
+    }
+
+    Set<Symbol> getNonTerminals() {
+        return nonterminals().collect(Collectors.toSet());
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="Algorithms">
+    boolean derivesToLambda(Symbol s, Stack<Tuple<Rule, Symbol>> recurseStack) {
```

Comment:
We can use ArrayDeque rather than Stack. The latter is essentially deprecated in favor of the former.

Reply:
Good call, I'll just make that change in master.

Reply:
Done: a710fb3e02